### PR TITLE
Add Page object

### PR DESCRIPTION
**Pull request by camillobruni**
Goal: Add a fixed interface to interact with test frames and their DOM nodes.
This will allow us to run more complex scenarios in the future without sacrificing flexibility or performance for the existing tests.

- Add `Page` object
- Refactor tests to interact with the frame only through the Page object

### CODE REVIEW

**Comment by camillobruni** [comment]
Depending on the resolution of PR #22 I might have to fix the mail client implementation to work with the `Page` API.

File: resources/benchmark-runner.mjs [hunk 1]
```diff
@@ -41,43 +41,59 @@ class BenchmarkState {
         return !this._testIndex;
     }
 
-    prepareCurrentSuite(runner, frame) {
+    async prepareCurrentSuite(frame, page) {
```
Comment (camillobruni) [inline resources/benchmark-runner.mjs:44]: The diff is still a bit messy due to depending on PR #19. 
For easier reviewing, I'll leave this PR as WIP until we land #19.

File: resources/benchmark-runner.mjs [hunk 2]
```diff
@@ -41,43 +41,71 @@ class BenchmarkState {
...
-            function resolveIfReady() {
-                const element = contentDocument.querySelector(selector);
-                if (element) {
-                    window.requestAnimationFrame(function () {
-                        return resolve(element);
-                    });
-                    return;
-                }
+          const resolveIfReady = () => {
+            const element = this.querySelector(selector);
```
Comment (rniwa) [inline resources/benchmark-runner.mjs:66]: Nit: What is up with the indentation here? Seems to be 2 space instead of 4?

File: resources/benchmark-runner.mjs [hunk 3]
```diff
@@ -41,43 +41,71 @@ class BenchmarkState {
...
                 setTimeout(resolveIfReady, 50);
             }
-
-            resolveIfReady();
+          };
+          resolveIfReady();
         });
     }
 
+    querySelector(selector) {
```
Comment (rniwa) [inline resources/benchmark-runner.mjs:79]: WebKit style is to put { in a new line.
I know we haven't been consistent but we might as well as in new code.

File: resources/benchmark-runner.mjs [hunk 4]
```diff
@@ -41,43 +41,71 @@ class BenchmarkState {
...
         });
     }
 }
 
-export class BenchmarkRunner {
-    constructor(suites, client) {
-        this._suites = suites;
-        this._prepareReturnValue = null;
-        this._client = client;
+const ENTER_KEY_CODE = 13;
```
Comment (rniwa) [inline resources/benchmark-runner.mjs:56]: It doesn't seem like this needs to be global?
We can just declare it within `triggerEnter` instead.

File: resources/benchmark-runner.mjs [hunk 5]
```diff
@@ -147,57 +172,54 @@ export class BenchmarkRunner {
         }, 0);
     }
 
-    step(state) {
+    async step(state) {
```
Comment (rniwa) [inline resources/benchmark-runner.mjs:175]: Can we do this `async` syntax adoption in a separate patch?
It's making the code review harder.

File: resources/benchmark-runner.mjs [hunk 6]
```diff
@@ -147,57 +172,54 @@ export class BenchmarkRunner {
...
         if (state.isFirstTest()) {
             this._removeFrame();
-            let self = this;
-            return state.prepareCurrentSuite(this, this._appendFrame()).then(prepareReturnValue => {
-                self._prepareReturnValue = prepareReturnValue;
-                return self._runTestAndRecordResults(state);
-            });
+            this._appendFrame();
+            this._page = new Page(this._frame);
+            await state.prepareCurrentSuite(this._frame, this._page);
```
Comment (rniwa) [inline resources/benchmark-runner.mjs:191]: Why are we passing both this._frame and this._page?
Can't we just use pass page here?

File: resources/benchmark-runner.mjs [hunk 7]
```diff
@@ -147,57 +172,54 @@ export class BenchmarkRunner {
...
         }
 
-        return this._runTestAndRecordResults(state);
+        return await this._runTestAndRecordResults(state);
     }
 
-    runAllSteps(startingState) {
+    async runAllSteps(startingState) {
```
Comment (rniwa) [inline resources/benchmark-runner.mjs:197]: Same comment about doing this `async` syntax adoption in a separate PR.

File: resources/tests.mjs [hunk 8]
```diff
@@ -4,39 +4,29 @@ const numberOfItemsToAdd = 100;
...
 Suites.push({
     name: 'VanillaJS-TodoMVC',
     url: 'todomvc/vanilla-examples/vanillajs/index.html',
-    prepare(runner, contentWindow, contentDocument) {
-        return runner.waitForElement('.new-todo').then(element => {
-            element.focus();
-            return element;
-        });
+    async prepare(page) {
+        const element = await page.waitForElement('.new-todo');
```
Comment (rniwa) [inline resources/tests.mjs:11]: This local variable seems unnecessary.
We can just do: `(await ~).focus();`

File: resources/tests.mjs [hunk 9]
```diff
@@ -106,31 +94,31 @@ Suites.push({
...
     name: 'React-TodoMVC',
     url: 'todomvc/architecture-examples/react/index.html',
-    prepare(runner, contentWindow, contentDocument) {
-        contentWindow.app.Utils.store = () => {};
-        return runner.waitForElement('.new-todo').then(element => {
-            element.focus();
-            return element;
-        });
+    async prepare(page) {
+        // FIXME: use app code
```
Comment (rniwa) [inline resources/tests.mjs:98]: What does this FIXME mean?

File: resources/tests.mjs [hunk 10]
```diff
@@ -227,28 +212,28 @@ Suites.push({
...
     url: 'todomvc/architecture-examples/backbone/index.html',
-    prepare(runner, contentWindow, contentDocument) {
-        contentWindow.Backbone.sync = () => {};
-        return runner.waitForElement('#appIsReady').then(element => {
-            const newTodo = contentDocument.querySelector('.new-todo');
-            newTodo.focus();
-            return newTodo;
-        });
+    async prepare(page) {
+        // FIXME: use app code
```
Comment (rniwa) [inline resources/tests.mjs:216]: Ditto.

File: resources/tests.mjs [hunk 11]
```diff
@@ -357,55 +339,53 @@ Suites.push({
...
+        new BenchmarkTestStep('CompletingAllItems', page => {
+            const checkboxes = page.querySelectorAll('.toggle');
             for (let i = 0; i < checkboxes.length; i++)
                 checkboxes[i].click();
         }),
-        new BenchmarkTestStep('DeletingAllItems', (newTodo, contentWindow, contentDocument) => {
+        new BenchmarkTestStep('DeletingAllItems', page => {
             for (let i = 0; i < numberOfItemsToAdd; i++)
-                contentDocument.querySelector('.destroy').click();
+            page.querySelector('.destroy').click();
```
Comment (rniwa) [inline resources/tests.mjs:362]: This seems like wrong indentation.

File: resources/tests.mjs [hunk 12]
```diff
@@ -415,37 +395,38 @@ Suites.push({
...
             for (let i = 0; i < numberOfItemsToAdd; i++)
-                contentDocument.querySelector('.destroy').click();
+                page.querySelector('.destroy').click();
         }),
     ]
 });
 
-function processElmWorkQueue(contentWindow) {
+function processElmWorkQueue(page) {
+    // TODO move workload to test
```
Comment (rniwa) [inline resources/tests.mjs:428]: What is this TODO about? Also, please use FIXME instead of TODO.

**Review by squarewave — COMMENTED** [review]
This all looks good to me, as far as the substance of the change goes. Thanks for taking this on.

File: resources/tests.mjs (hunk 8 above)
Comment (camillobruni) [reply resources/tests.mjs:11]: DONE.

**Comment by camillobruni** [comment]
I've addressed your comments.
Will upload the async-conversion PR soon.

File: resources/benchmark-runner.mjs [hunk 13]
```diff
@@ -5,24 +5,29 @@ export class BenchmarkTestStep {
     }
 }
 
-class BenchmarkState {
-    constructor(suites) {
+class BenchmarkState
```
Comment (rniwa) [inline resources/benchmark-runner.mjs:8]: Oh, class declaration puts { on the same line. Just function declaration that puts { on a new line.
Sorry, should have been more clear.

File: resources/tests.mjs [hunk 14]
```diff
@@ -415,37 +394,38 @@ Suites.push({
...
             for (let i = 0; i < checkboxes.length; i++)
                 checkboxes[i].click();
         }),
-        new BenchmarkTestStep('DeletingItems', (params, contentWindow, contentDocument) => {
-            const deleteButtons = contentDocument.querySelectorAll('.destroy');
+        new BenchmarkTestStep('DeletingItems', page => {
+            const deleteButtons = page.querySelectorAll('.destroy');
             for (let i = 0; i < numberOfItemsToAdd; i++)
-                contentDocument.querySelector('.destroy').click();
+                page.querySelector('.destroy').click();
```
Comment (rniwa) [inline resources/tests.mjs:421]: Huh, it seems unnecessary to keep calling querySelector if we to all the buttons to delete in the first place.
This is unrelated to your change so we can fix it up in a separate PR.

File: resources/tests.mjs [hunk 15]
```diff
@@ -415,37 +394,38 @@ Suites.push({
...
         }),
     ]
 });
 
-function processElmWorkQueue(contentWindow) {
+function processElmWorkQueue(page) {
```
Comment (rniwa) [inline resources/tests.mjs:426]: Nit: { should be on new line.

**Review by rniwa — APPROVED** [review]
r=me

**Review by bgrins — APPROVED** [review]
Approving as per Doug's comment

**Comment by bgrins** [comment]
@camillobruni is this good to land?

**Comment by camillobruni** [comment]
Sorry, long weekend. double checking now.
